Backport ES2K changes to ipdk-io

ffoulkes opened the pull request:
Open-source ES2K support

nupuruttarwar committed:
Initial code for MEV support

Signed-off-by: Nupur Uttarwar <nupur.uttarwar@intel.com>

nupuruttarwar committed:
Add config file for ES2000

Signed-off-by: Nupur Uttarwar <nupur.uttarwar@intel.com>

nupuruttarwar committed:
Add port config file and remove unsupported functionality for ES2000

Signed-off-by: Nupur Uttarwar <nupur.uttarwar@intel.com>

nupuruttarwar committed:
Disable bf_shell and enable logging under /var/log/stratum

Signed-off-by: Nupur Uttarwar <nupur.uttarwar@intel.com>

ffoulkes committed:
Merge pull request #1 from intel-innersource/mev_supp

Initial code for MEV support

n-sandeep committed:
Changes in stratum to rename from ES2000 to ES2K

Signed-off-by: Sandeep N <sandeep.nagapattinam@intel.com>

ffoulkes committed:
Merge pull request #2 from intel-innersource/rename_es2000_to_es2k

Changes in stratum to rename from ES2000 to ES2K

5abeel committed:
Update common.proto with IPsec enums & messages

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
IPsec  gnmi_events and notification helper functions

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Update Copyright date

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Adding Fixed Function manager and update sde_interface sde_wrapper files

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Update tdi_node with fixed_function_manager

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Update Credentials Manager to sync with latest upstream

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Enable TLS-mode

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Update Copyright date

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Adding IPsec manager files

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Add yang parse tree files for ipsec

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Update Es2k main, hal and switch and IPsecManager

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Update gnmi_publisher with ipsec handling

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Update tls mode and es2k support

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Update check for symlinks & credentials_manager to stringent security checks

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

weiqiand committed:
porting flow dump chnages to mev

Signed-off-by: weiqiand <weiqian.dai@intel.com>

Koushik-kopanati committed:
Merge pull request #6 from intel-innersource/jan25_flowdump

porting flow dump changes to mev

5abeel committed:
Update es2ksw ptr in ipsec yang parse tree

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Fix SET messages to go through SwitchInterface

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Fix DELETE messages to go through SwitchInterface

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

n-sandeep committed:
Merge branch '23_01_dpdk_stratum' into HEAD

n-sandeep committed:
Merge pull request #7 from intel-innersource/merge_from_dpdk_main

Merge from dpdk main

5abeel committed:
Update GET messages to go through SwitchInterface

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Cleanup unused code

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Fix initialization and issues found during test

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Rebase + resolve conflicts

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

ffoulkes committed:
Ported krnlmon synchronization changes to ES2K (#10)

Signed-off-by: Derek G Foster <derek.foster@intel.com>

5abeel committed:
Update and merging conflicts

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Merge pull request #3 from intel-innersource/ipsec_support

IPsec support

ffoulkes committed:
Remove target-specific #include from common file (#13)

- Removed an errant #include of an ES2K header file from a common file.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

bharticemk committed:
adding support for mirror in stratum for splitArchitecture

Signed-off-by: bharticemk <richa.bharti@intel.com>

bharticemk committed:
Merge pull request #12 from intel-innersource/mirror_splitArch

adding support for mirror in stratum for splitArchitecture

5abeel committed:
Addressing coverity issues

Signed-off-by: 5abeel <sabeel.ansari@intel.com>

5abeel committed:
Merge pull request #15 from intel-innersource/coverity_fixes

Addressing coverity issues

5abeel committed:
Fix for IPsec encryption key format to use ietf-yang:hex-string format

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Remove commented line

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Merge pull request #17 from intel-innersource/ipsec_enc_key_format

Fix for IPsec encryption key format to use ietf-yang:hex-string format

5abeel committed:
Address Coverity reported warnings to remove deadcode

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Merge pull request #18 from intel-innersource/remove-deadcode

Address Coverity reported warnings to remove deadcode

5abeel committed:
common.proto changes with more descriptive comments + typo fix

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
common.proto changes with more descriptive comments + typo fix

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Change message name from LifetimeHard to HardLifetime

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Change message name from LifetimeHard to HardLifetime

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

ffoulkes committed:
Fix breaking changes in stratum (#19)

* Fix breaking change in TdiNode

- Fixed breaking change in TdiNode::CreateInstance by removing the
  unused TdiFixedFunctionManager object pointer.

- Streamlined error handling in Read/Write ExternEntry methods
  (and made it more consistent) by using a common error return
  for unsupported entry types.

* Address breaking changes in TdiSdeWrapper class

- A block of target-specific #includes was added to the target-
  agnostic tdi_sde_interface.h file. They have been removed,
  and the single #include that is actually needed has been
  added conditionally to tdi_wrapper.cc.

- A target-specific ipsec notification method was added to the
  target-agnostic tdi_sde_wrapper class. As a temporary workaround,
  it has been conditionalized using ES2K_TARGET. The method needs
  to be reengineered for compatibility with the TDI architecture.

- Corrected target type in a couple of comments

Signed-off-by: Derek G Foster <derek.foster@intel.com>

---------

Signed-off-by: Derek G Foster <derek.foster@intel.com>

5abeel committed:
Move ConvertIPsecNotificationToString from common/utils to yang_parse_tree_ipsec to keep it local

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Move ConvertIPsecNotificationToString from common/utils to yang_parse_tree_ipsec to keep it local

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Update names to match canonical spelling from Ipsec to IPsec

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
merge commit

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Merge pull request #20 from intel-innersource/ipsec_support

IPsec changes to address comments previously raised

ffoulkes committed:
Extract es2k_port_manager from es2k_sde_target (#22)

- When the port management methods were separated from TdiSdeWrapper,
  the implementations were modified in place. This changelist moves the
  port methods from es2k_sde_target.cc to a new es2k_port_manager.cc
  file, paralleling the existing es2k_port_manager.h file.

Signed-off-by: Derek Foster <derek.foster@intel.com>

ffoulkes committed:
Fix format warning in tdi_ipsec_manager (#23)

- Changed %d to %lu in printf format string to match argument type.

Signed-off-by: Derek Foster <derek.foster@intel.com>

ffoulkes committed:
Update Bazel build and add ES2K support (#24)

* Update Bazel build and add ES2K support

Broad-spectrum update to restore the ability to build Stratum and
run unit tests using Bazel. Incorporated a few changes from the
parent project. Added support for ES2K.

This changelist should be sufficient to support unit testing.
Additional work is required to complete ES2K support.

- Updated .bazelversion to 3.5.0 to 3.7.2. This tracks an upstream
  change.

- Implemented external repository for ES2K SDE.

- Defined targets for tdi_lut_manager, tdi_fixed_function_manager,
  tdi_ipsec_manager, and es2k_port_manager.

- Updated compiler flags (defs.bzl) to track upstream changes.

- Updated workspace_rule to track upstream changes. This allows us
  to specify sha256 hashes for remote workspaces.

- Updated dependencies to track upstream changes. Changed grpc
  (v1.32.2 -> v1.40.0), googleapis, p4lang/p4c, nlohmann json
  (v3.9.1 -> v3.10.4), abseil-cpp (20210324 -> 20220623), and
  rules_go (v0.20.3 -> v0.24.11). Also added sha256 hashes
  where available.

- Overrode the definition of p4runtime to fetch the repository
  from ipdk-io instead of p4lang. THIS CHANGE MUST BE REMOVED
  BEFORE WE CAN UPSTREAM TO THE PARENT PROJECT.

- Addressed link errors by adding missing libraries to external
  repositories for the DPDK and Tofino SDEs.

- Removed call to helper_log_always_return_true() from the
  RET_CHECK macro, to address a compilation issue. This tracks
  an upstream change.

- Fixed compiler warnings in tdi_ipsec_manager.cc and
  tdi_sde_wrapper.cc

- Updated tdi_sde_mock with new methods. Added mock for the
  TdiPortManager class.

- Supplied some missing dependencies.

* Fix issues in TdiSdeMock class

- Added missing SetExact() method in TableKeyInterface.

- Defined notification_table_callback_t typedef to specify the
  signature of the notification table callback function. Used it
  to mock the InitNotificationTableWithCallback method.

* Work around link issue with stratum-es2k target

- Manually specify linker parameters for glib-2.0.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Implement DpdkPortManager class (#25)

- Extract dpdk_port_manager.h from tdi_sde_interface.h and
  tdi_sde_wrapper.h. Include the PortConfigParams and
  HotplugConfigParams classes, which are specific to DPDK.

- Extract dpdk_port_manager.cc from dpdk_sde_target.cc.

- Update DpdkChassisManager constructor and factory function to
  accept DpdkPortManager* in place of TdiSdeInterface*.

- Update Bazel build files for dpdk_port_manager

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Finish implementing TofinoPortManager class (#26)

- Extract tofino_port_manager.cc from tofino_sde_target.cc.

- Declare methods in tofino_port_manager.h to be virtual, so they can be overridden by the TofinoPortManagerMock.

- Implement TofinoPortManagerMock class.

- Update TofinoChassisManagerTest to work with TofinoPortManager.

- Update Bazel build files.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Stratum unit test updates (#27)

- Create separate library and header file targets in SDEs to provide
  better granularity in build target dependencies. In particular, this
  allows a target to depend on just the header files.

- Provide target_defines for config_monitoring_service and p4_service,
  which contain DPDK_TARGET #ifdefs.

- Disable symlink check in config_monitoring_service, which breaks
  several of the Stratum unit tests. This change needs to be
  reconsidered.

- Make several //stratum/hal/lib/tdi header targets dependent on
  target_sdk_headers rather than target_sdk. Introduce a new
  tdi_sde_wrapper_interface target.

- Remove redundant parse_tree_mock target and correct sole dependency
  to reference yang_parse_tree_mock target instead.

- Fix a compiler error in procmon_test.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Repair unit tests (#29)

- Create mocks for the DpdkPortManager, Es2kPortManager, and
  IPsecManager classes.

- Revise dpdk_chassis_manager_test to interact with DpdkPortManager
  instead of TdiSdeWrapper. Remove tests for features not supported
  by DpdkChassisManager.

- Remove unused TdiSdeInterface pointer from Es2kChassisManager.

- Revise es2k_chassis_manager_test to interact with Es2kPortManager
  instead of TdiSdeWrapper, and eliminate the unused PhalInterface
  pointer. Delete or disable tests for features not supported by
  Es2kChassisManager.

- Make Es2kPortManager methods virtual, so they can be overridden
  by Es2kPortManagerMock.

- Update es2k_switch_test to pass IPsecManagerMock object to
  Es2kSwitch factory method. Remove unmatched #endif. Eliminate
  unused TdiSdeMock object.

- Refactor the `incompatible_enable_tdi_legacy_bytestream_responses`
  flag, moving the definition to tdi_sde_flags.cc and the declaration
  to tdi_sde_flags.h. This eliminates the duplicate definitions and
  declarations, and makes it possible to depend on the definition
  (rather than the SDE) in unit tests that require the flag.

- Remove unused is_model definition from es2k_main.cc.

- Update Bazel build files. Make link options for glib-2.0 part of the
  ES2K SDE library definition. Define RTE library targets for the DPDK
  and ES2K SDEs. Supply missing dependencies.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

5abeel committed:
Remove log statement (#30)

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

5abeel committed:
Move target specific code from target agnostic codebase (#32)

This moves the IPsec init notification registration from target-agnostic to target-specific codebase

The header file contents still remain in target-agnostic code -- will have to figure out how to address those later

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

ffoulkes committed:
Address unit test dependency issues (#31)

- Implement port manager dummy classes to reduce runtime dependencies
  on the SDE in the switch and chassis manager unit tests.

- Remove unused Bazel tofino_defines and target_sdk definitions.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Move TdiPortManager to its own header file (#33)

- Extract TdiPortManager and PortStatusEvent class definitions from
  the TdiSdeInterface class and move them to their own header file.
  Update references and dependencies as appropriate.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

aashishkuma committed:
tdi_lut_manager.cc UT

aashishkuma committed:
Addressing few comments

ffoulkes committed:
Rename IPsecManager class to TdiIpsecManager (#34)

- Rename the IPsecManager class to conform to Stratum naming
  and capitalization conventions.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

aashishkuma committed:
UT for tdi_action_profile_manager.cc

aashishkuma committed:
Delete tdi_lut_manager_test.cc

Deleting the file as this file UT has been raised as part of separate PR.

aashishkuma committed:
Update BUILD

aashishkuma committed:
tdi_lut_manager.cc UT (#35)

Implement unit test for `TdiLutManager` class.

ffoulkes committed:
Fix bazel test error on CI server (#39)

Problem: The CI server fails with the Python exception
`AttributeError: module 'collections' has no attribute 'MutableMapping'`
when running Stratum unit tests under Bazel.

MutableMapping was moved from 'collections' to 'collections.abc' in
Python 3.3. Support for the old import was deprecated, and it was
announced that it would be removed after Python 3.9. The CI server
is running Python 3.10.

On examination, we found that:
1) The code containing the old import is in rules_python.
2) Our version of Stratum uses (pre)release 0.0.1 of rules_python.
   The latest release of rules_python is 0.21.0.
3) The upstream version of Stratum has eliminated rules_python
   entirely.

We have addressed the issue by merging changes from upstream
to eliminate rules_python from our repository as well.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Minor gnmi_publisher edits (#40)

- Update formatting to conform to coding standard.

- Supply missing copyright notices.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

aashishkuma committed:
Addressing comments

aashishkuma committed:
Merge branch 'aaskumar/ActionProfMgrUT' of https://github.com/intel-innersource/networking.ipu.mountevans.network-recipes.stratum-dev into aaskumar/ActionProfMgrUT

aashishkuma committed:
Addressing comments

aashishkuma committed:
Addressing comments

aashishkuma committed:
Merge pull request #38 from intel-innersource/aaskumar/ActionProfMgrUT

TdiActionProfileManager unit test

n-sandeep committed:
Create ES2K node specific to target

Changes are made to
- Split ES2K node from TDI node, this is to support any specific changes
  to ES2K target and not break the existing TDI node which is shared by
  DPDK and Tofino.
- Any PNA specific implementations can be added to this node without
  overlapping TNA specifications used by Tofino and available in TDI
  node.
- Example:
  With mirroring feature added for PNA, we intoduced few more externs
  like MATCH_LOOKUP_TABLE and TERNARY_MATCH_LOOKUP_TABLE which are
  overlapping with tofino TDI specific externs like
  TnaExternActionProfileId and TnaExternActionSelectorId

Signed-off-by: Sandeep N <sandeep.nagapattinam@intel.com>

n-sandeep committed:
Remove duplicate definitions from derived class Es2kNode

1) Remove unwanted definitions in ES2K node and re-use the same from
base class TdiNode.
2) Definitions that re retained are
- WriteForwardingEntries
- ReadForwardingEntries
- WriteExternEntry
- ReadExternEntry
3) Initialize base class TdiNode members from Es2kNode

Signed-off-by: Sandeep N <sandeep.nagapattinam@intel.com>

aashishkuma committed:
TdiFixedFunctionManagerUT

n-sandeep committed:
Merge pull request #41 from intel-innersource/introduce_es2k_node

Create ES2K node specific to target

aashishkuma committed:
Addressed Comments

aashishkuma committed:
Adding failure test cases

ffoulkes committed:
Standardize TDI Main interface (#43)

This CL provides a common interface for starting Stratum.
It eliminates all the #ifdefs in infrap4d_main (which is a Good Thing).

- Rename the DpdkMain, Es2kMain, and TofinoMain functions to Main,
  and define a common interface header file. This simplifies infrap4d_main
  or any other module that needs to start Stratum.

  Note that the Tofino build does not support krnlmon (by design),
  and hence does not implement the the extended interface used by
  infrap4d_main.

- Use buildifier to canonicalize some of the Bazel BUILD files.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

aashishkuma committed:
Addressed Comments

aashishkuma committed:
Merge pull request #42 from intel-innersource/aaskumar/tdiFixedFuncMgrUT

TdiFixedFunctionManagerUT

nupuruttarwar commented on stratum/pipelines/ptf/scapy_exec.bzl:
Are we removing files under stratum/pipelines/ since latest stratum doesn't have this? 
Are we rebasing the code to latest stratum as well as part of this PR

nupuruttarwar commented on stratum/lib/utils.cc:
Where are we using this function?
I understood from Derek that this is to transport the encryption key via Protobuf for IPSEC. If it's specific to how we transport the key for ES2ky, then we should introduce a functionsn in utils module under tdi instead of modifying common stratum lib utils

nupuruttarwar commented on stratum/hal/lib/yang/yang_parse_tree_paths.h:
for IPSEC?

nupuruttarwar commented on stratum/hal/lib/yang/yang_parse_tree.h:
will be executed in case node is deleted?

nupuruttarwar commented on stratum/hal/lib/tdi/tdi_lut_manager.cc:
line 65 and 69 can be removed

nupuruttarwar commented on stratum/hal/lib/tdi/tdi_constants.h:
Will changing these values affect backward compatibility?

nupuruttarwar commented on stratum/hal/bin/tdi/es2k/es2k_skip_p4.conf:
changes to support e2100 value are already merged, we can change this to e2100

nupuruttarwar commented on stratum/hal/lib/tdi/es2k/es2k_port_manager.cc:
We don't add port for ES2K. we can keep the function and return OK. Function on line 146 can be removed for ES2K

ffoulkes commented on stratum/lib/utils.cc:
This function is only used in `TdiIpsecManager`. I'd suggest that the function be defined there.

5abeel reviewed (approved):
LGTM

ffoulkes commented on stratum/pipelines/ptf/scapy_exec.bzl:
Fixing the Bazel build required, among other things, that we update the version of gRPC used in the Bazel build. This broke other parts of the build. The most straightforward solution to the cascade of issues was to cherry-pick the dependencies and their supporting changes from the latest version of Stratum.

One of the issues in the cascade was that we could no longer build the Python files under `stratum/pipelines`. The latest version of stratum removed these files and their BUILD rules, so I did likewise. Problem solved, and we are one step closer to updating to the latest version of stratum.

We are not rebasing the code to the latest version of Stratum. That's still in our future.

ffoulkes commented on stratum/hal/bin/tdi/es2k/es2k_skip_p4.conf:
The instructions from Dan are to wait until after the merge.

ffoulkes committed:
Fix es2k_node breakage (#75)

The implementation of the Es2kNode class did not include updating
the Bazel build or the Stratum unit tests that were affected by the
change. This CL addresses the omissions.

- Fix compile errors in es2k_node.cc and es2k_node.h.

- Remove logic to make tdi_lut_manager an optional parameter. This
  was only necessary to allow TdiNode to support non-ES2K targets.
  Move it to a more appropriate position in the parameter list.

- Reorder methods in es2k_node.h to match the order in which they
  appear in tdi_node.h.

- Define es2k_node_mock class. Update es2k_switch_test to use
  Es2kNodeMock instead of TdiNodeMock. Disable tests that are
  broken because the attributes they test are unimplemented.

- Update Bazel BUILD files to support the above changes.

Signed-off-by: Derek G Foster <derek.foster@intel.com>
